### PR TITLE
JS: Add insufficient key size query

### PR DESCRIPTION
**Pull request by erik-krogh**
[Evaluation](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/cwe326-eval/reports) looks fine.  
It finds a few results among the test-cases of NodeJS.   

Heavily inspired by: [csharp query](https://github.com/github/codeql/blob/main/csharp/ql/src/Security%20Features/InsufficientKeySize.ql), [go query](https://github.com/github/codeql-go/blob/main/ql/src/Security/CWE-326/InsufficientKeySize.ql), [python query](https://github.com/github/codeql/blob/main/python/ql/src/Security/CWE-326/WeakCryptoKey.ql). 

There aren't that many projects that generate their own keys, but I managed to [find a few](https://lgtm.com/query/8376072927259659767/).   
None of them look that serious, so I've set the severity to `warning`.

### CODE REVIEW

**Review by Misran28 — APPROVED** [review]

**Review by RasmusWL — COMMENTED** [review]
I'm not 100% sure about the change to `python/ql/lib/semmle/python/concepts/CryptoAlgorithms.qll`. Treating AES-CBC as secure got me a little suspicious, and doing a bit of research revealed that indeed, CBC can be vulnerable to a _padding oracle attack_ (described in blogposts at [Microsoft](https://docs.microsoft.com/en-us/dotnet/standard/security/vulnerabilities-cbc-mode) and [Cloudflare](https://blog.cloudflare.com/padding-oracles-and-the-decline-of-cbc-mode-ciphersuites/)). I'm not 100% sure how we should deal with this, but very interested to hear if you had thought about this :blush: 

In regards to the new query, I can see a difference with the one we have in Python. This query covers symmetric and asymmetric algorithms, and the Python query covers asymmetric (specifically RSA/DSA) and elliptic curve cryptography (ECC). So maybe you should look at adding ECC for this query? (and maybe we should look at adding support for symmetric algorithms in the Python version)

**Review by esbena — COMMENTED** [review]
LGTM.

The qhelp lacks examples, but the language versions of this query also lack those, so I suppose that is alright.

**Review by Misran28 — APPROVED** [review]

**Comment by erik-krogh** [comment]
> I'm not 100% sure about the change to `python/ql/lib/semmle/python/concepts/CryptoAlgorithms.qll`. Treating AES-CBC as secure got me a little suspicious, and doing a bit of research revealed that indeed, CBC can be vulnerable to a _padding oracle attack_ (described in blogposts at [Microsoft](https://docs.microsoft.com/en-us/dotnet/standard/security/vulnerabilities-cbc-mode) and [Cloudflare](https://blog.cloudflare.com/padding-oracles-and-the-decline-of-cbc-mode-ciphersuites/)). I'm not 100% sure how we should deal with this, but very interested to hear if you had thought about this blush

Yes, CBC is bad (TIL), but I don't think that's something we can handle in `CryptoAlgorithms.qll`.   
The encryption algorithm used (AES) is considered safe (with a strong enough key), and the attack against CBC seems to require a pretty stable timing side-channel. 

I think hardcoding some mode-of-operations and only adding them to a few of the algorithms was the wrong approach.  
I've instead changed the `matchesName` predicate such that it ignores everything after a dash. 
That way modes-of-operations can be included in the name, without us having to explicitly handle it. 

> In regards to the new query, I can see a difference with the one we have in Python. This query covers symmetric and asymmetric algorithms, and the Python query covers asymmetric (specifically RSA/DSA) and elliptic curve cryptography (ECC). So maybe you should look at adding ECC for this query? (and maybe we should look at adding support for symmetric algorithms in the Python version)

I just added the what I found to be relevant for the libraries I encountered.   
E.g. [the NodeJS crypto library method for creating an ECC object](https://nodejs.org/api/crypto.html#cryptocreateecdhcurvename) does not allow explicitly setting the length of the key, so I didn't see a need to model ECC.

**Review by esbena — APPROVED** [review]

File: python/ql/lib/semmle/python/concepts/CryptoAlgorithms.qll [hunk 1]
```diff
@@ -85,11 +85,12 @@ abstract class CryptographicAlgorithm extends TCryptographicAlgorithm {
 
   /**
    * Holds if the name of this algorithm matches `name` modulo case,
-   * white space, dashes, and underscores.
+   * white space, dashes, underscores, and anything after a dash in the name.
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/concepts/CryptoAlgorithms.qll:88]:
```suggestion
   * white space, dashes, underscores, and anything after a dash in the name
   * (to ignore modes of operation, such as CBC or ECB).
```

File: javascript/ql/lib/semmle/javascript/security/CryptoAlgorithms.qll [hunk 2]
```diff
@@ -85,11 +85,12 @@ abstract class CryptographicAlgorithm extends TCryptographicAlgorithm {
 
   /**
    * Holds if the name of this algorithm matches `name` modulo case,
-   * white space, dashes, and underscores.
+   * white space, dashes, underscores, and anything after a dash in the name.
```
Comment (RasmusWL) [inline javascript/ql/lib/semmle/javascript/security/CryptoAlgorithms.qll:88]:
```suggestion
   * white space, dashes, underscores, and anything after a dash in the name
   * (to ignore modes of operation, such as CBC or ECB).
```

**Review by RasmusWL — CHANGES_REQUESTED** [review]
Thanks for looking into the modes of operation. I think we could probably handle this better, but for now I guess it looks ok :+1:

I would like a small comment explaining _why_ we ignore everything after the dash, but otherwise LGTM :+1:

**Review by RasmusWL — APPROVED** [review]

**Comment by erik-krogh** [comment]
@esbena are you OK with the changes? Can I get a re-approve?

**Review by esbena — APPROVED** [review]

**Pull request merged**